Python/bytecodes.c

```diff
@@ -2286,38 +2286,40 @@ dummy_func(
         }
 
         op(_LOAD_ATTR_WITH_HINT, (hint/1, owner, dict: PyDictObject * -- attr)) {
+            PyDictKeysObject *dk = FT_ATOMIC_LOAD_PTR(dict->ma_keys);
             PyObject *attr_o;
-            if (!LOCK_OBJECT(dict)) {
-                POP_INPUT(dict);
-                DEOPT_IF(true);
-            }
-
-            if (hint >= (size_t)dict->ma_keys->dk_nentries) {
-                UNLOCK_OBJECT(dict);
+            if (hint >= (size_t)dk->dk_nentries) {
                 POP_INPUT(dict);
                 DEOPT_IF(true);
             }
             PyObject *name = GETITEM(FRAME_CO_NAMES, oparg>>1);
-            if (dict->ma_keys->dk_kind != DICT_KEYS_UNICODE) {
-                UNLOCK_OBJECT(dict);
+            if (dk->dk_kind != DICT_KEYS_UNICODE) {
                 POP_INPUT(dict);
                 DEOPT_IF(true);
             }
-            PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dict->ma_keys) + hint;
+            PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dk) + hint;
             if (ep->me_key != name) {
-                UNLOCK_OBJECT(dict);
                 POP_INPUT(dict);
                 DEOPT_IF(true);
             }
-            attr_o = ep->me_value;
+            attr_o = FT_ATOMIC_LOAD_PTR(ep->me_value);
             if (attr_o == NULL) {
-                UNLOCK_OBJECT(dict);
                 POP_INPUT(dict);
                 DEOPT_IF(true);
             }
-            STAT_INC(LOAD_ATTR, hit);
+#ifdef Py_GIL_DISABLED
+            if (!_Py_TryIncrefCompareStackRef(&ep->me_value, attr_o, &attr)) {
+                POP_INPUT(dict);
+                DEOPT_IF(true);
+            } else if (dk != FT_ATOMIC_LOAD_PTR(dict->ma_keys)) {
+                PyStackRef_CLOSE(attr);
+                POP_INPUT(dict);
+                DEOPT_IF(true);
+            }
```

Comment:
I don't think we need this check `dk != FT_ATOMIC_LOAD_PTR(dict->ma_keys)`. Even if the `dict->ma_keys` changed, we should still have loaded a valid previous value at this point.

Reply:
That seems right to me, I think that means we also don't need the check here? https://github.com/python/cpython/blob/main/Objects/dictobject.c#L1527

Reply:
Yeah. There's also a similar check a few lines below that in the same function:

https://github.com/python/cpython/blob/69426fcee7fcecbe34be66d2c5b58b6d0ffe2809/Objects/dictobject.c#L1547-L1550